### PR TITLE
heketi: add support for profiling with pprof

### LINKED ISSUE #799: how to use pprof to profile heketi binaries
Are there instructions how to profile heketi binaries with pprof? As far as I can see in heketi code pprof is not included

### PR DESCRIPTION
**Pull request by nixpanic**
### What does this PR achieve? Why do we need it?

Add support for profiling like "net/http/pprof" does, but through the
"gorilla/mux" router. Profiling is disabled by default, set the
`HEKETI_PROFILING` environment variable to a non-empty string or add
`"profiling": true` to the --config file.

Several endpoints will get registered, and are writter to stderr:

    $ HEKETI_PROFILING=1 ./heketi --config=heketi.json
    ...
    DEBUG: registered profiling handler on /debug/pprof/block
    DEBUG: registered profiling handler on /debug/pprof/goroutine
    DEBUG: registered profiling handler on /debug/pprof/heap
    DEBUG: registered profiling handler on /debug/pprof/mutex
    DEBUG: registered profiling handler on /debug/pprof/threadcreate
    DEBUG: registered profiling handler on /debug/pprof/cmdline
    DEBUG: registered profiling handler on /debug/pprof/profile
    DEBUG: registered profiling handler on /debug/pprof/symbol
    DEBUG: registered profiling handler on /debug/pprof/trace
    ...

Standard tools like `go tool pprof` and `curl` work. Examples:

    $ go tool pprof http://localhost:8080/debug/pprof/profile
    $ go tool pprof http://localhost:8080/debug/pprof/heap
    $ curl http://localhost:8080/debug/pprof/heap?debug=1

### Does this PR fix issues?

Fixes #799

### Notes for the reviewer

https://golang.org/pkg/net/http/pprof/ contains more details about the pprof interface.

### CODE REVIEW

File: pkg/utils/profiling.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+// General Public License, version 3 or any later version (LGPLv3 or
+// later), or the GNU General Public License, version 2 (GPLv2), in all
+// cases as published by the Free Software Foundation.
+
+package utils
+
+// Based on ideas from https://github.com/mistifyio/negroni-pprof
+
+import (
+	"github.com/gorilla/mux"
```
Comment (Madhu-1) [inline pkg/utils/profiling.go]: @nixpanic import order needs to be updated.

**Comment by nixpanic** [comment]
Updated based on last comments from John.

https://hackmd.io/8RLNb_bQSWOJbbGOGaexcA?view# contains a start for how to use this feature.

File: server/profiling/profiling.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+	}
+
+	// static profiles as listed in net/http/pprof/pprof.go:init()
+	addPath(router, "cmdline", http.HandlerFunc(pprof.Cmdline))
+	addPath(router, "profile", http.HandlerFunc(pprof.Profile))
+	addPath(router, "symbol", http.HandlerFunc(pprof.Symbol))
+	addPath(router, "trace", http.HandlerFunc(pprof.Trace))
+
+	// The index handler only lists the runtime_pprof.Profiles(), so let's not index any.
+	//router.Methods("GET").Path(basePath).Name("index").Handler(http.HandlerFunc(pprof.Index))
```
Comment (phlogistonjohn) [inline server/profiling/profiling.go]: Should we drop this commented out line?

**Comment by phlogistonjohn** [comment]
That document looks great. Since you've already written it can you please make it part of the PR as a document in docs directory?
I promise not to be picky about the file name as long as you add a caveat to the top that the feature is primarily intended for developers.
Beyond that I'm pretty happy with it and look forward to merging it.

File: server/profiling/profiling.go (hunk 2 above)
Comment (nixpanic) [reply server/profiling/profiling.go]: I thought about that, but I prefer to keep it as that is enabled by default for other ways of integrating pprof. When someone wants to add this, it is clear that not having it is intentional.

Comment (phlogistonjohn) [reply server/profiling/profiling.go]: Er, can't we just have the comment saying that? Something like:
```
// The index handler only lists the runtime_pprof.Profiles() which does not include all the endpoints
// we have. Thus we opt not to use the common prof.Index endpoint here
```

Comment (nixpanic) [reply server/profiling/profiling.go]: ok, done!

**Comment by nixpanic** [comment]
@heketi/maintainers I think this is ready now.

**Pull request merged**